nats/aio/client.py

```diff
@@ -72,29 +72,169 @@
+        self._max_msgs = max_msgs
+        self._received = 0
+        self._cb = cb
+        self._iterate = iterate
+        self._future = future
 
         # Per subscription message processor
-        self.pending_msgs_limit = None
-        self.pending_bytes_limit = None
-        self.pending_queue = None
-        self.pending_size = 0
-        self.wait_for_msgs_task = None
+        self._pending_msgs_limit = pending_msgs_limit
+        self._pending_bytes_limit = pending_bytes_limit
+        self._pending_queue = asyncio.Queue(maxsize=pending_msgs_limit)
+        self._pending_size = 0
+        self._wait_for_msgs_task = None
+
+        self._message_iterator = None
+        if self._iterate:
+            self._message_iterator = _SubscriptionMessageIterator(
+                self._pending_queue
+            )
+
+    @property
+    def messages(self):
+        if not self._iterate:
+            raise NatsError(
+                "cannot iterate over messages with a non iteration subscription type"
+            )
+
+        return self._message_iterator
+
+    def _start(self, error_cb):
+        if self._cb:
+            if not asyncio.iscoroutinefunction(self._cb) and \
+                not (hasattr(self._cb, "func") and asyncio.iscoroutinefunction(self._cb.func)):
+                raise NatsError("nats: must use coroutine for subscriptions")
+
+            self._wait_for_msgs_task = asyncio.create_task(
+                self._wait_for_msgs(error_cb)
+            )
+
+        elif self._iterate:
+            # Used with async for
+            pass
+        elif self._future:
+            # Used to handle the single response from a request.
+            pass
+        else:
+            raise NatsError("nats: invalid subscription type")
+
+    async def drain(self):
+        try:
+            # Announce server that no longer want to receive more
+            # messages in this sub and just process the ones remaining.
+            await self._conn._send_unsubscribe(self._id)
+
+            # Roundtrip to ensure that the server has sent all messages.
+            await self._conn.flush()
+
+            if self._pending_queue:
+                # Wait until no more messages are left,
+                # then cancel the subscription task.
+                await self._pending_queue.join()
+
+            # stop waiting for messages
+            self._stop_processing()
+
+            # Subscription is done and won't be receiving further
+            # messages so can throw it away now.
+            self._conn._remove_sub(self._id)
+        except asyncio.CancelledError:
+            # In case draining of a connection times out then
+            # the sub per task will be canceled as well.
+            pass
+
+    async def unsubscribe(self, limit=0):
+        if self._conn.is_closed:
+            raise ErrConnectionClosed
+        if self._conn.is_draining:
+            raise ErrConnectionDraining
+
+        self._stop_processing()
```

Comment:
I think here we need a special case for when `limit` is 0 so that it can replace `auto_unsubscribe` usage like:

```python
    # Stop receiving after 2 messages.
    sub = await nc.subscribe("foo", cb=message_handler)
    await sub.unsubscribe(limit=2)
    await nc.publish("foo", b'Hello')
    await nc.publish("foo", b'World')
    await nc.publish("foo", b'Not delivered')
    await nc.flush()
```

Then the 3rd message would not be delivered.

```suggestion
        if limit == 0:
                self._stop_processing()
```